File: backend/app/schemas/pdf.py
from datetime import datetime

from app.models.enums import DependencyType, DocumentProcessingStatus, ExtractionMethod
from app.schemas.common import ApiModel


class CurriculumDocumentResponse(ApiModel):
    id: str
    programa_id: str
    nombre_archivo: str
    tipo_archivo: str
    estado_procesamiento: DocumentProcessingStatus
    porcentaje_progreso: int
    fecha_carga: datetime


class DocumentExtractionResponse(ApiModel):
    id: str
    documento_malla_id: str
    texto_extraido: str
    metodo_extraccion: ExtractionMethod
    confianza_ocr: float
    fecha_procesamiento: datetime


class DetectedCourseResponse(ApiModel):
    codigo: str
    nombre: str
    creditos: int
    semestre: int
    confianza: float


class DetectedDependencyResponse(ApiModel):
    materia: str
    requiere: str
    tipo: DependencyType
    confianza: float


class PdfProcessingDiagnostics(ApiModel):
    page_count: int
    pages_with_native_text: int
    pages_using_ocr: int
    pages_without_text: int
    scanned_like: bool
    ocr_available: bool
    ocr_language_used: str | None = None
    can_retry_with_ocr: bool
    recommended_action: str
    message: str


class PdfProcessResponse(ApiModel):
    document: CurriculumDocumentResponse
    extraction: DocumentExtractionResponse
    courses: list[DetectedCourseResponse]
    dependencies: list[DetectedDependencyResponse]
    diagnostics: PdfProcessingDiagnostics


class OcrStatusResponse(ApiModel):
    available: bool
    engine: str | None = None
    installed_path: str | None = None
    languages: list[str]
    spanish_language_available: bool
    ready_for_scanned_pdfs: bool
    issues: list[str]
    next_steps: list[str]
    message: str


class ApproveGraphRequest(ApiModel):
    version_id: str
    courses: list[DetectedCourseResponse]
    dependencies: list[DetectedDependencyResponse]


class ApproveGraphResponse(ApiModel):
    document_id: str
    approved: bool
    created_courses: int
    updated_courses: int
    created_dependencies: int
    message: str

File: src/types/pdf.ts
export type ProcessingStatus =
  | 'pendiente'
  | 'extrayendo_texto'
  | 'ocr'
  | 'procesando'
  | 'validando'
  | 'convertido_a_grafo'
  | 'error'

export interface DocumentoMalla {
  id: string
  programaId: string
  nombreArchivo: string
  tipoArchivo: string
  estadoProcesamiento: ProcessingStatus
  porcentajeProgreso: number
  fechaCarga: string
}

export interface ExtraccionDocumento {
  id: string
  documentoMallaId: string
  textoExtraido: string
  metodoExtraccion: 'texto_pdf' | 'ocr_imagen' | 'mixto'
  confianzaOcr: number
  fechaProcesamiento: string
}

export interface ChunkDocumento {
  id: string
  documentoMallaId: string
  contenido: string
  orden: number
  fuente: string
}

export interface DetectedCourse {
  codigo: string
  nombre: string
  creditos: number
  semestre: number
  confianza: number
}

export interface DetectedDependency {
  materia: string
  requiere: string
  tipo: 'prerequisito' | 'correquisito'
  confianza: number
}

export interface PdfProcessingDiagnostics {
  pageCount: number
  pagesWithNativeText: number
  pagesUsingOcr: number
  pagesWithoutText: number
  scannedLike: boolean
  ocrAvailable: boolean
  ocrLanguageUsed?: string
  canRetryWithOcr: boolean
  recommendedAction: 'review' | 'manual_review' | 'install_ocr_and_retry'
  message: string
}

export interface OcrStatus {
  available: boolean
  engine?: string
  installedPath?: string
  languages: string[]
  spanishLanguageAvailable: boolean
  readyForScannedPdfs: boolean
  issues: string[]
  nextSteps: string[]
  message: string
}
